Automattic/woocommerce-payments
# Update UPE logic around the cards payment method and improve onboarding styling

Fixes #4114 

#### Changes proposed in this Pull Request

**Add active and locked card payments method when UPE is active + styling improvements**

The initial issue (#4114) surfaced the fact that a merchant could not complete the UPE onboarding _without_ selecting one of the specific UPE payment methods even though it was not a requirement since the cards payment method would suffice.

The Stripe UPE logic imposes that the cards payment method be mandatory, but our UI didn't reflect that reality. I decided to adjust _both_ the UPE onboarding wizard and the payments settings view to be consistent with this requirement. Previously, the cards payment checkbox list item was not present in the wizard, and the cards checkbox control in the settings view was not disabled when UPE was in use.

This PR fixes these logical and UX inconsistencies by **forcing the cards payment method checkbox to be checked and disabled** both in the wizard and the settings when UPE is active. To let the user know why this behavior is present, a `Required` label is added next to the cards payment method name, in both views.

At the same time, to help the user be better informed about requirements and possible next steps, **the wizard button label changes** from `Add payment methods` to `Continue` to capture better the multiple scenarios a user might face (either selecting UPE-specific payment methods or just sticking with card payments and test things). To ease the possible fear of users who are just testing, we **introduce a description under the payment methods selection box** with the mention: `You can always change or add more payment methods later.`

As [discussed](https://github.com/Automattic/woocommerce-payments/issues/4114#issuecomment-1344101862) with @rogermattic, all these changes bring _consistency and predictability_ to users who want to experiment with the UPE early access.

Throughout the initial exploration of the issue, several design inconsistencies surfaced. It seemed like a good moment to steer the current state of the UPE onboarding wizard closer to the original design intent. As a result, this PR introduces **styling changes around spacing, font weight and sizes, and alignments.** We are not at pixel-perfect fidelity (that was not my intention), but we are closer and can convey that increased level of quality that builds trust with our merchants.

The initial state of the UPE onboarding wizard
![Markup on 2022-12-15 at 22:04:11](https://user-images.githubusercontent.com/8830539/207956209-62928b91-7953-4c87-b74f-6e68e2fa893d.png)


The updated state of the UPE onboarding wizard
![Markup on 2022-12-15 at 22:01:39](https://user-images.githubusercontent.com/8830539/207955804-c6c2d346-76de-4979-81a7-f8eb66b8e841.png)


The Payments Settings view after the changes
<img width="1079" alt="Markup 2022-12-15 at 21 57 21" src="https://user-images.githubusercontent.com/8830539/207955244-7cea2f17-2cde-44bb-aa57-b92950090fa4.png">

#### Dev notes

In surfacing the cards payment method list item in the UPE onboarding wizard, I explored simply adding the `card` entry to the UPE payment methods list and letting the rest of the logic just display it: this _does_ make sense since the card payment method _is_ a valid (and required) UPE payment method, but it would have touched _a pretty large surface of the code;_ I understood that the UPE payment methods list is treated as _UPE-specific_ or _UPE-additional_ payment methods, so I left it as such.

In the end, I chose to _hard-code_ the cards payment method list item at the top since it is in line with the Stripe UPE requirements.

I had to introduce the behavior of `required` and/or `locked` payment methods. I chose to separate a payment method being required from it being locked (its input control disabled) to offer more future flexibility to our components in dealing with either combination. This behavioral addition is made at the general level, in our components.

Through minor refactoring of some variables, I aim to clarify that a payment method which is determined to be disabled (due to UPE-specific logic) doesn't necessarily refer to its input control (checkbox in our case) having the `disabled` HTML attribute.

In terms of CSS styling changes and additions, I aimed to keep my rules specific to avoid the risk of accidental overflows into other areas. There might be an argument to be made to make the styling more reusable, but I opted to be defensive and reduce the styling specificity when the need arises.

#### Testing instructions

1. On a dev site with WooCommerce Payments, checkout this PR's branch: `update/4114-upe-onboarding-with-just-credit-cards-method`
2. Don't enable additional UPE payment methods and finish the KYC flow to enable your account, or ensure that "Enable UPE checkout" is not checked in the **WCPay Dev Utils**.
3. Go to **Payments > Settings**.
4. The `Credit card / debit card` payment method should not be required, and you can activate or deactivate it at will.
5. Deactivate the `Credit card / debit card` payment method.
6. Click the button label "Enable in your store" to go to the UPE onboarding wizard.
7. In the wizard, the `Credit card / debit card` payment method should be checked, required, and without the possibility of unchecking it; at the same time, the "Continue" button should be active and allow you to finish your UPE onboarding without the need to select any UPE-specific payment methods.
8. Click the "Continue" button, then "Go to payments settings" button to return to the Settings view.
9. The `Credit card / debit card` payment method should be checked, required, and without the possibility of unchecking it.
10. Disable UPE from the three-dots right-side dropdown and the subsequent modal.
11. The `Credit card / debit card` payment method should remain active but no longer required or with the checkbox disabled.

Here is a quick rundown through the steps above:
![Screen Capture on 2022-12-15 at 22-35-34](https://user-images.githubusercontent.com/8830539/207963749-26cb1997-59c7-487c-a9a9-71ba6f14ca1c.gif)

-------------------


**Post merge**
